### PR TITLE
Add error messages and unit tests for setups where min_budget >= max_budget

### PR DESCRIPTION
**Pull request by Bronzila**
Before we simply used an assertion and printed:
```
Only (Max Budget > Min Budget) is supported for DEHB.
```

Now we log an error with same message if ```min_budget <= max_budget```, but we also add an additional error if ```min_budget == max_budget```, guiding the user to use plain DE (values are obviously filled with values specified by the user, apart from configspace and target_function):

```
2023-07-25 16:54:19.802 | ERROR    | src.dehb.optimizers.dehb:__init__:75 - Only (Max Budget > Min Budget) is supported for DEHB.
2023-07-25 16:54:19.803 | ERROR    | src.dehb.optimizers.dehb:__init__:77 - If you have a fixed fidelity, you can instead run DE as follows: AsyncDE(cs=configspace, f=target_function, dimensions=1, pop_size=2, max_age=inf, mutation_factor=0.5, crossover_prob=0.5, strategy=rand1_bin, budget=27, boundary_fix_type=random)
```

### CODE REVIEW

File: src/dehb/optimizers/dehb.py [hunk 1]
```diff
@@ -59,7 +71,19 @@ def __init__(self, cs=None, f=None, dimensions=None, mutation_factor=None,
...
+            if self.max_budget == self.min_budget:
+                self.logger.error(
+                    "If you have a fixed fidelity, " \
+                    "you can instead run DE as follows: " \
+                    "AsyncDE(cs=configspace, f=target_function, dimensions=" \
+                    f"{self.dimensions}, pop_size={self.dimensions * 2}, " \
+                    f"max_age={self.max_age}, mutation_factor=" \
+                    f"{self.mutation_factor}, crossover_prob={self.crossover_prob}, "\
+                    f"strategy={self.strategy}, budget={self.max_budget}, " \
+                    f"boundary_fix_type={self.fix_type})")
```
Comment (Neeratyoy) [inline src/dehb/optimizers/dehb.py:85]: Maybe we just point to the DE documentation (when ready)?
Right now, this is probably too verbose, don't you think?
We could also just direct the user to the `DE/AsyncDE` class.

File: src/dehb/optimizers/dehb.py [hunk 2]
```diff
@@ -59,7 +71,19 @@ def __init__(self, cs=None, f=None, dimensions=None, mutation_factor=None,
...
+            sys.exit()
```
Comment (Neeratyoy) [inline src/dehb/optimizers/dehb.py:86]: We should double-check when this happens such that the termination here does not affect the Dask processes.

File: tests/test_dehb.py [hunk 3]
```diff
@@ -98,3 +97,11 @@ def test_brackets_exhaustion(self):
...
         assert dehb._is_run_budget_exhausted(brackets=1), "Run budget should be exhausted"
 
+class TestInitialization:
+    """Class that bundles all tests regarding the initialization of DEHB."""
+    def test_higher_min_budget(self):
+        """Test that verifies, that DEHB breaks if min_budget > max_budget."""
+        cs = create_toy_searchspace()
+        with pytest.raises(SystemExit):
+            create_toy_optimizer(configspace=cs, min_budget=28, max_budget=27, eta=3,
+                                        objective_function=objective_function)
```
Comment (Neeratyoy) [inline tests/test_dehb.py:107]: This will also be updated based on the final fate of the `sys.exit` above.

**Review by Neeratyoy — CHANGES_REQUESTED** [review]
Looks good. Maybe one more iteration and we can merge.

File: src/dehb/optimizers/dehb.py (hunk 2 above)
Comment (Bronzila) [reply src/dehb/optimizers/dehb.py:86]: This only happens when the initializer of the DEHBBase class is called, so before any Dask clients are set or initialized. Therefore I don't think this should be a problem. However I thought about rather raising an ```AsssertionError``` to be a little more specific than just exiting. Under the hood this obviously results in the same outcome.

File: src/dehb/optimizers/dehb.py [hunk 4]
```diff
@@ -59,7 +71,14 @@ def __init__(self, cs=None, f=None, dimensions=None, mutation_factor=None,
...
         self.max_budget = max_budget
-        assert self.max_budget > self.min_budget, "only (Max Budget > Min Budget) supported!"
+        if self.max_budget <= self.min_budget:
+            self.logger.error("Only (Max Budget > Min Budget) is supported for DEHB.")
+            if self.max_budget == self.min_budget:
+                self.logger.error(
+                    "If you have a fixed fidelity, " \
+                    "you can instead run DE. For more information checkout: " \
+                    "https://automl.github.io/DEHB/references/de")
+            raise AssertionError()
```
Comment (Neeratyoy) [inline src/dehb/optimizers/dehb.py:81]: 👍🏽

File: tests/test_dehb.py [hunk 5]
```diff
@@ -98,3 +97,11 @@ def test_brackets_exhaustion(self):
 
         assert dehb._is_run_budget_exhausted(brackets=1), "Run budget should be exhausted"
 
+class TestInitialization:
+    """Class that bundles all tests regarding the initialization of DEHB."""
+    def test_higher_min_budget(self):
+        """Test that verifies, that DEHB breaks if min_budget > max_budget."""
+        cs = create_toy_searchspace()
+        with pytest.raises(AssertionError):
+            create_toy_optimizer(configspace=cs, min_budget=28, max_budget=27, eta=3,
```
Comment (Neeratyoy) [inline tests/test_dehb.py:106]: also maybe an equality test to be sure, as that is a more legitimate case

File: src/dehb/optimizers/dehb.py [hunk 6]
```diff
@@ -27,6 +27,18 @@ def __init__(self, cs=None, f=None, dimensions=None, mutation_factor=None,
...
+        os.makedirs(self.output_path, exist_ok=True)
+        self.logger = logger
+        log_suffix = time.strftime("%x %X %Z")
+        log_suffix = log_suffix.replace("/", '-').replace(":", '-').replace(" ", '_')
+        self.logger.add(
+            "{}/dehb_{}.log".format(self.output_path, log_suffix),
+            **_logger_props
+        )
+        self.log_filename = "{}/dehb_{}.log".format(self.output_path, log_suffix)
+
```
Comment (Neeratyoy) [inline src/dehb/optimizers/dehb.py:41]: can we have this as a separate function such as `_setup_logger()` or equivalent to keep the `__init__` as clutter-free as possible?

**Comment by Bronzila** [comment]
Just checked: both using ```n_workers``` and passing a dask ```Client``` is safe. I think this can be merged now. @Neeratyoy

File: src/dehb/optimizers/dehb.py (hunk 6 above)
Comment (Neeratyoy) [reply src/dehb/optimizers/dehb.py]: Can we move this outside `__init__` and call it in?
For only aesthetic reasons.

**Review by Neeratyoy — APPROVED** [review]

**Comment by Neeratyoy** [comment]
I leave the timing of the branch cleaning part to you for now. Thanks!

**Pull request merged**